### PR TITLE
[FEATURE F1.1.3] Type Conversion Layer — Tensor ↔ Atom Bridge

### LINKED ISSUE #78: [FEATURE F1.1.3] Type Conversion Layer
## Feature: F1.1.3

**Parent Phase:** 1.1
**Parent Epic:** E1
**Priority:** High
**Effort:** 5 points

### Description

Convert between neural tensors and symbolic atoms

### Acceptance Criteria

- [ ] Feature implemented
- [ ] Tests passing
- [ ] Documentation updated

---
*Generated from Unreal-Echo Development Roadmap*
*Feature ID: F1.1.3 | Phase: 1.1 | Epic: E1*

### PR DESCRIPTION
**Pull request by drzo**
## Summary

Implements the Type Conversion Layer (F1.1.3) — bidirectional conversion between **Eigen neural tensors** and **OpenCog AtomSpace atoms**. This bridges the neural-symbolic gap in the Deep Tree Echo cognitive architecture.

Closes #78

### Pure C++23 library in `OpenCogEcho/include/opencog/bridge/`

**Tensor → Atom (4 converters):**
| Function | Input | Output |
|----------|-------|--------|
| `activation_to_nodes()` | `VectorXf` | `ConceptNode` set with `TruthValue{activation, confidence}` |
| `matrix_to_links()` | `MatrixXf` | `EvaluationLink` graph (pred + src/tgt nodes) |
| `state_to_atom()` | `VectorXf` | `StateLink(AnchorNode, ListLink(NumberNode...))` |
| `temporal_pattern_to_atoms()` | `MatrixXf` (time × features) | `AtTimeLink` chain |

**Atom → Tensor (4 converters):**
| Function | Input | Output |
|----------|-------|--------|
| `nodes_to_activation()` | `ConceptNode` set | `VectorXf` (TV strength per node) |
| `links_to_matrix()` | `EvaluationLink` graph | `MatrixXf` |
| `atom_to_state()` | `StateLink` | `VectorXf` |
| `attention_focus_to_vector()` | Attentional focus set | `VectorXf` (STI values) |

**`TypeConversionBridge` class** — unified API with:
- Round-trip validation (`validate_roundtrip_activation()`, `validate_roundtrip_nodes()`)
- Batch conversion
- Configurable `ConversionPolicy` (threshold, normalization, naming, presets)
- Performance metrics tracking

## Test Results

**34 new tests, all passing** (674 total in OpenCogEcho suite, 0 failures)

Categories: tensor→atom (13), atom→tensor (6), round-trip (3), bridge API (4), edge cases (5), policies (3)

## Files Changed

```
OpenCogEcho/
├── include/opencog/bridge/
│   ├── tensor_atom_types.hpp           # Shared types, policies, ConversionResult<T>
│   ├── tensor_to_atom.hpp              # Neural → Symbolic API
│   ├── atom_to_tensor.hpp              # Symbolic → Neural API
│   └── type_conversion_bridge.hpp      # High-level bridge class
├── src/bridge/
│   ├── tensor_to_atom.cpp
│   ├── atom_to_tensor.cpp
│   └── type_conversion_bridge.cpp
├── tests/
│   └── test_type_conversion_bridge.cpp # 34 tests
└── CMakeLists.txt                      # +bridge sources, +Eigen dep, +test
```

Also updated: `FEATURE_F1.1.3_IMPLEMENTATION_SUMMARY.md`

## Dependencies

- OpenCog core types (already in-tree)
- Eigen 3.x (already in-tree)
- **No Unreal Engine dependency** — fully testable standalone

---

> [!NOTE]
> **Medium Risk**
> New neural–symbolic integration mutates AtomSpace (including state replacement/orphan cleanup) and becomes a public dependency via Eigen; risk is mitigated by extensive tests but errors could affect downstream cognitive pipelines.
> 
> **Overview**
> Introduces **Feature F1.1.3: Type Conversion Layer** under `OpenCogEcho/include/opencog/bridge/` and `src/bridge/`, complementing the existing message transport with **data transformation** between neural tensors and symbolic atoms.
> 
> **Tensor → atom:** `activation_to_nodes`, `matrix_to_links`, `state_to_atom`, and `temporal_pattern_to_atoms` map `VectorXf`/`MatrixXf` into `ConceptNode`, `EvaluationLink`, `StateLink`, and `AtTimeLink` structures with configurable thresholding, normalization, TruthValue/AttentionValue mapping, and NaN/Inf guards.
> 
> **Atom → tensor:** `nodes_to_activation`, `links_to_matrix`, `atom_to_state`, and `attention_focus_to_vector` rebuild Eigen vectors/matrices from atom graphs and attention focus.
> 
> **`TypeConversionBridge`** wraps both directions with `ConversionPolicy` presets, batch activation conversion, latency/error **metrics**, and **round-trip validation** helpers.
> 
> **Build & docs:** `OpenCogEcho/CMakeLists.txt` compiles the three bridge `.cpp` files, links **Eigen3::Eigen** publicly, and registers `test_type_conversion_bridge.cpp` when headers exist. `FEATURE_F1.1.3_IMPLEMENTATION_SUMMARY.md` is updated to document this layer instead of the prior bidirectional message-protocol write-up.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 5316337325603099fd8cbd2b09bed517590e0f7a. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: OpenCogEcho/src/bridge/tensor_to_atom.cpp [hunk 1]
```diff
@@ -0,0 +1,288 @@
...
+        float strength = std::clamp(val, 0.0f, 1.0f);
+        float confidence = derive_confidence(val, policy);
+
+        Handle h = space.add_node(AtomType::CONCEPT_NODE, name,
+                                  TruthValue{strength, confidence});
+
+        if (policy.map_activation_to_sti) {
+            AttentionValue av{val * policy.sti_scale, 0, 0};
+            space.set_av(h, av);
+        }
```
[the remark this answers is not in the document]
Comment (drzo) [reply OpenCogEcho/src/bridge/tensor_to_atom.cpp:119]: Fixed in 8231908. Added explicit \set_tv()\ calls after every \dd_node()\/\dd_link()\ in \ctivation_to_nodes\, \matrix_to_links\, and \	emporal_pattern_to_atoms\. This ensures the TruthValue is always updated even when the AtomTable returns an existing deduplicated atom.

File: OpenCogEcho/src/bridge/type_conversion_bridge.cpp [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+    const std::vector<Eigen::VectorXf>& activations)
+{
+    std::vector<ConversionResult<std::vector<Handle>>> results;
+    results.reserve(activations.size());
+
+    for (auto& vec : activations) {
+        results.push_back(convert_activation(vec));
+    }
+    return results;
+}
```
[the remark this answers is not in the document]
Comment (drzo) [reply OpenCogEcho/src/bridge/type_conversion_bridge.cpp:202]: Fixed in 8231908. batch_convert_activations now generates a unique prefix per batch entry (appending _batchN to the policy prefix), preventing AtomTable name collisions between different activation vectors.

File: OpenCogEcho/src/bridge/tensor_to_atom.cpp [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+    if (auto err = validate_vector(state); err != ConversionError::None)
+        return ConversionResult<Handle>::fail(err);
+
+    // Create number nodes for each element
+    std::vector<Handle> elements;
+    elements.reserve(static_cast<size_t>(state.size()));
+    for (int i = 0; i < state.size(); ++i) {
+        std::string num_name = std::to_string(state[i]);
+        Handle num = space.add_node(AtomType::NUMBER_NODE, num_name);
+        elements.push_back(num);
```
[the remark this answers is not in the document]
Comment (drzo) [reply OpenCogEcho/src/bridge/tensor_to_atom.cpp:211]: Fixed in 8231908. Replaced std::to_string(float) with std::to_chars for full-precision float serialization. Round-trip test tolerance tightened from 1e-4 to 1e-6.

File: OpenCogEcho/src/bridge/tensor_to_atom.cpp [hunk 4]
```diff
@@ -0,0 +1,288 @@
...
+    std::vector<AtomId> ids;
+    ids.reserve(elements.size());
+    for (auto& h : elements) ids.push_back(h.id());
+    Handle list = space.add_link(AtomType::LIST_LINK, std::span<const AtomId>{ids});
+
+    // Anchor for the state
+    Handle anchor = space.add_node(AtomType::ANCHOR_NODE, label);
+
+    // StateLink: (StateLink anchor list)
+    Handle state_link = space.add_link(AtomType::STATE_LINK, {anchor, list});
```
[the remark this answers is not in the document]
Comment (drzo) [reply OpenCogEcho/src/bridge/tensor_to_atom.cpp]: Fixed in 8231908. state_to_atom now removes any existing StateLink for the anchor before creating the new one, maintaining single-state-per-key semantics.

File: OpenCogEcho/src/bridge/tensor_to_atom.cpp [hunk 5]
```diff
@@ -0,0 +1,314 @@
...
+
+    // Anchor for the state
+    Handle anchor = space.add_node(AtomType::ANCHOR_NODE, label);
+
+    // Remove any previous StateLink for this anchor so we don't accumulate
+    // stale state atoms (StateLink semantics expect one state per key).
+    auto incoming = space.get_incoming_by_type(anchor, AtomType::STATE_LINK);
+    for (auto& old_link : incoming) {
+        space.remove(old_link, false);
+    }
```
[the remark this answers is not in the document]
Comment (drzo) [reply OpenCogEcho/src/bridge/tensor_to_atom.cpp:251]: Fixed in 839bf9e. The old StateLink removal now also cleans up orphaned ListLink and NumberNode children — each is removed only if it has no remaining incoming references, preventing unbounded memory growth during ESN streaming.

File: OpenCogEcho/src/bridge/tensor_to_atom.cpp [hunk 6]
```diff
@@ -0,0 +1,333 @@
...
+            if (space.get_incoming(old_list).empty()) {
+                space.remove(old_list, false);
+            }
+            // Remove NumberNodes that are no longer referenced
+            for (auto& num : old_nums) {
+                if (space.get_incoming(num).empty()) {
+                    space.remove(num, false);
+                }
+            }
+        }
```
[the remark this answers is not in the document]
Comment (drzo) [reply OpenCogEcho/src/bridge/tensor_to_atom.cpp:250]: Fixed in 531633732. Reordered to create the new StateLink *before* cleaning up old ones. Shared ListLink/NumberNode atoms from deduplication are now protected by the new StateLink's incoming reference, so the orphan check correctly keeps them alive.
